File: cpp/437. Path Sum III.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// Prefixsum, Time: O(n), Space: O(n)
// Concept is same as LeetCode 560
public:
    void dfs(TreeNode *root, int runningSum, int sum, unordered_map<int, int> &counter, int &ans) {
        if(!root) return;
        runningSum += root->val;
        if(runningSum == sum) ans++;
        if(counter.count(runningSum-sum)) ans += counter[runningSum-sum];
        counter[runningSum]++;
        dfs(root->left, runningSum, sum, counter, ans);
        dfs(root->right, runningSum, sum, counter, ans);
        counter[runningSum]--;
    }
    int pathSum(TreeNode* root, int sum) {
        unordered_map<int, int> counter;
        int ans = 0;
        dfs(root, 0, sum, counter, ans);
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    TreeNode *lll = new TreeNode(3);
    TreeNode *llr = new TreeNode(-2);
    TreeNode *ll = new TreeNode(3, lll, llr);
    TreeNode *lrr = new TreeNode(1);
    TreeNode *lr = new TreeNode(2, nullptr, lrr);
    TreeNode *l = new TreeNode(5, ll, lr);
    TreeNode *rr = new TreeNode(11);
    TreeNode *r = new TreeNode(-3, nullptr, rr);
    TreeNode *root = new TreeNode(10, l, r);
    Solution solution;
    int ans = solution.pathSum(root, 8);
    assert(ans==3);
}

File: python/437-Path Sum III.py
from bintree_helper import *

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        "2 nested DFS, Time: O(n^2)"
        if root is None: return 0
        def dfs(root, k):
            "returns number of paths add up to k from root include root"
            count = 0
            if root.val == k: count += 1
            if root.left: count += dfs(root.left, k-root.val)
            if root.right: count += dfs(root.right, k-root.val)
            return count
        return dfs(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)
    def pathSum(self, root: TreeNode, sum: int) -> int:
        "Prefixsum, Time: O(n), Space: O(n)"
        "Concept is same as LeetCode 560"
        "https://www.youtube.com/watch?v=uZzvivFkgtM"
        if root is None: return 0
        from collections import Counter
        result = 0
        prefixsum = Counter()
        def dfs(root, running_sum):
            nonlocal result
            running_sum += root.val
            if running_sum == sum: result += 1
            result += prefixsum[running_sum-sum]
            prefixsum[running_sum] += 1
            if root.left:
                dfs(root.left, running_sum)
            if root.right:
                dfs(root.right, running_sum)
            prefixsum[running_sum] -= 1
        dfs(root, 0)
        return result

root = deserialize("[10,5,-3,3,2,null,11,3,-2,null,1]")
sum = 8
result = Solution().pathSum(root, sum)
assert result == 3

root = deserialize("[1,-2,-3,1,3,-2,null,-1]")
sum = -1
result = Solution().pathSum(root, sum)
assert result == 4